test/integration/ctl/ctl_create_cr_test.go

```diff
@@ -0,0 +1,211 @@
+			inputArgs:      []string{cr1Name},
+			inputNamespace: ns1,
+			keyFilename:    "test.key",
+			expErr:         false,
+			expNamespace:   ns1,
+			expName:        cr1Name,
+			expKeyFilename: "test.key",
+		},
+	}
+
+	for name, test := range tests {
+		// Run ctl create cr command with input options
+		t.Run(name, func(t *testing.T) {
+			streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+			// Options to run create CR command
+			opts := &certificaterequest.Options{
+				CMClient:         cmCl,
+				RESTConfig:       config,
+				IOStreams:        streams,
+				CmdNamespace:     test.inputNamespace,
+				EnforceNamespace: test.inputNamespace != "",
+				KeyFilename:      test.keyFilename,
+			}
+
+			opts.InputFilenames = []string{test.inputFile}
+
+			err := opts.Run(test.inputArgs)
+
+			if err != nil {
+				if !test.expErr {
+					t.Errorf("got unexpected error when trying to create CR: %v", err)
+				}
+				t.Logf("got an error, which was expected, details: %v", err)
+				return
+			} else {
+				// got no error
+				if test.expErr {
+					t.Errorf("expected but got no error when creating CR")
+				}
+			}
+
+			// Finished creating CR, check if everything is expected
+			crName := test.inputArgs[0]
+			gotCr, err := cmCl.CertmanagerV1alpha2().CertificateRequests(test.inputNamespace).Get(ctx, crName, metav1.GetOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if gotCr.Namespace != test.expNamespace {
+				cmCl.CertmanagerV1alpha2().CertificateRequests(gotCr.Namespace).Delete(ctx, gotCr.Name, metav1.DeleteOptions{})
+				t.Errorf("CR created in unexpected Namespace, expected: %s, actual: %s", test.expNamespace, gotCr.Namespace)
+			}
+
+			if gotCr.Name != test.expName {
+				cmCl.CertmanagerV1alpha2().CertificateRequests(gotCr.Namespace).Delete(ctx, gotCr.Name, metav1.DeleteOptions{})
+				t.Errorf("CR created has unexpected Name, expectedL %s, actualL %s", test.expName, gotCr.Name)
+			}
+
+			// Check the file where the private key is stored
+			keyData, err := ioutil.ReadFile(test.expKeyFilename)
+			if err != nil {
+				cmCl.CertmanagerV1alpha2().CertificateRequests(gotCr.Namespace).Delete(ctx, gotCr.Name, metav1.DeleteOptions{})
+				t.Errorf("error when reading file storing private key: %v", err)
+			}
+			_, err = pki.DecodePrivateKeyBytes(keyData)
+			if err != nil {
+				cmCl.CertmanagerV1alpha2().CertificateRequests(gotCr.Namespace).Delete(ctx, gotCr.Name, metav1.DeleteOptions{})
+				t.Errorf("invalid private key: %v", err)
+			}
+
+			// Clean up CertificateRequest
+			// Everything is expected, so clean up with what is expected
+			err = cmCl.CertmanagerV1alpha2().CertificateRequests(test.expNamespace).Delete(ctx, test.expName, metav1.DeleteOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+
+	// Clean up tmp folder with private key files
+	if err := os.Chdir(".."); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.RemoveAll("tmp"); err != nil {
+		t.Fatal(err)
+	}
```

Comment:
Not sure what’s going on here, but we should be using TempDir to create a temporary directory, and using ‘defer’ immediately afterward to take care of deleting that directory after the test has completed 🙂

Reply:
If I use os.TempDir(), it tries to create `/var/folders/yy/rsf02qj9139gwvs14pnq90d00000gn/T/` and fails because it already exists.

Reply:
TempDir in the `os` package returns the global temporary directory: https://golang.org/pkg/os/#TempDir

There is another `TempDir` function in `ioutil` which can be used to create a *new* temporary directory 😅  https://yourbasic.org/golang/temporary-file-directory/

taken from that page:

```go
dir, err := ioutil.TempDir("dir", "prefix")
if err != nil {
	log.Fatal(err)
}
defer os.RemoveAll(dir)
```